// @netlify/vite-plugin-react-router/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name2 in all)
    __defProp(target, name2, { get: all[name2], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  createRequestHandler: () => createRequestHandler,
  default: () => netlifyPlugin,
  netlifyRouterContext: () => netlifyRouterContext
});
module.exports = __toCommonJS(index_exports);

// src/lib/handler.ts
var import_react_router = require("react-router");
function createNetlifyRequestHandler({
  build,
  mode,
  getLoadContext,
  netlifyRouterContext: netlifyRouterContext2,
  runtimeName
}) {
  const reactRouterHandler = (0, import_react_router.createRequestHandler)(build, mode);
  return async (request, netlifyContext) => {
    const start = Date.now();
    console.log(`[${request.method}] ${request.url}`);
    try {
      const getDefaultReactRouterContext = () => {
        const ctx = new import_react_router.RouterContextProvider();
        ctx.set(netlifyRouterContext2, netlifyContext);
        Object.assign(ctx, netlifyContext);
        return ctx;
      };
      const reactRouterContext = await getLoadContext?.(request, netlifyContext) ?? getDefaultReactRouterContext();
      const response = await reactRouterHandler(request, reactRouterContext);
      response.headers.set("x-nf-runtime", runtimeName);
      console.log(`[${response.status}] ${request.url} (${Date.now() - start}ms)`);
      return response;
    } catch (error) {
      console.error(error);
      return new Response("Internal Error", { status: 500 });
    }
  };
}

// src/lib/context.ts
var import_react_router2 = require("react-router");
function createNetlifyRouterContext() {
  return (0, import_react_router2.createContext)(
    new Proxy(
      // Can't reference `Netlify.context` here because it isn't set outside of a request lifecycle
      {},
      {
        get(_target, prop, receiver) {
          return Reflect.get(Netlify.context ?? {}, prop, receiver);
        },
        set(_target, prop, value, receiver) {
          return Reflect.set(Netlify.context ?? {}, prop, value, receiver);
        },
        has(_target, prop) {
          return Reflect.has(Netlify.context ?? {}, prop);
        },
        deleteProperty(_target, prop) {
          return Reflect.deleteProperty(Netlify.context ?? {}, prop);
        },
        ownKeys(_target) {
          return Reflect.ownKeys(Netlify.context ?? {});
        },
        getOwnPropertyDescriptor(_target, prop) {
          return Reflect.getOwnPropertyDescriptor(Netlify.context ?? {}, prop);
        }
      }
    )
  );
}

// src/runtimes/netlify-functions.ts
var netlifyRouterContext = createNetlifyRouterContext();
function createRequestHandler({
  build,
  mode,
  getLoadContext
}) {
  return createNetlifyRequestHandler({
    build,
    mode,
    getLoadContext,
    netlifyRouterContext,
    runtimeName: "Node"
  });
}

// src/plugin.ts
var import_promises = require("fs/promises");
var import_node_path2 = require("path");
var import_posix = require("path/posix");
var import_node_fetch_server = require("@remix-run/node-fetch-server");
var import_tinyglobby = require("tinyglobby");

// package.json
var name = "@netlify/vite-plugin-react-router";
var version = "3.1.1";

// src/lib/rollup.ts
var import_node_path = require("path");
var getAliasName = (id) => {
  const base = (0, import_node_path.basename)(id);
  return base.slice(0, Math.max(0, base.length - (0, import_node_path.extname)(id).length));
};
var normalizeRollupInput = (input) => {
  if (input == null) {
    return {};
  }
  if (typeof input === "string") {
    return { [getAliasName(input)]: input };
  }
  if (Array.isArray(input)) {
    return Object.fromEntries(input.map((entry) => [getAliasName(entry), entry]));
  }
  return input;
};
var mergeRollupInput = (existing, newEntries) => {
  return {
    ...normalizeRollupInput(existing),
    ...newEntries
  };
};

// src/plugin.ts
var NETLIFY_FUNCTIONS_DIR = ".netlify/v1/functions";
var NETLIFY_EDGE_FUNCTIONS_DIR = ".netlify/v1/edge-functions";
var FUNCTION_FILENAME = "react-router-server.mjs";
var FUNCTION_HANDLER_CHUNK = "server";
var FUNCTION_HANDLER_MODULE_ID = "virtual:netlify-server";
var RESOLVED_FUNCTION_HANDLER_MODULE_ID = `\0${FUNCTION_HANDLER_MODULE_ID}`;
var SERVER_ENTRY_MODULE_ID = "virtual:netlify-server-entry";
var toPosixPath = (path) => path.split(import_node_path2.sep).join(import_posix.sep);
var ALLOWED_USER_EDGE_FUNCTION_HANDLER_FILENAMES = [
  "server.ts",
  "server.mts",
  "server.cts",
  "server.mjs",
  "server.cjs",
  "server.js"
];
var findUserEdgeFunctionHandlerFile = async (root) => {
  for (const filename of ALLOWED_USER_EDGE_FUNCTION_HANDLER_FILENAMES) {
    try {
      await (0, import_promises.access)((0, import_node_path2.join)(root, filename));
      return filename;
    } catch {
    }
  }
  throw new Error(
    "Your Hydrogen site must include a `server.ts` (or js/mjs/cjs/mts/cts) file at the root to deploy to Netlify. See https://github.com/netlify/hydrogen-template."
  );
};
var FUNCTION_HANDLER = (
  /* js */
  `
import { createRequestHandler } from "@netlify/vite-plugin-react-router/serverless";
import * as build from "virtual:react-router/server-build";
export default createRequestHandler({
  build,
});
`
);
var EDGE_FUNCTION_HANDLER = (
  /* js */
  `
import { createRequestHandler } from "@netlify/vite-plugin-react-router/edge";
import * as build from "virtual:react-router/server-build";
export default createRequestHandler({
  build,
});
`
);
function generateNetlifyFunction(handlerPath, excludedPath) {
  return (
    /* js */
    `
    export { default } from "${handlerPath}";

    export const config = {
      name: "React Router server handler",
      generator: "${name}@${version}",
      path: "/*",
      excludedPath: ${JSON.stringify(excludedPath)},
      preferStatic: true,
    };
    `
  );
}
function generateEdgeFunction(handlerPath, excludedPath) {
  return (
    /* js */
    `
    export { default } from "${handlerPath}";

    export const config = {
      name: "React Router server handler",
      generator: "${name}@${version}",
      cache: "manual",
      path: "/*",
      excludedPath: ${JSON.stringify(excludedPath)},
    };
    `
  );
}
function netlifyPlugin(options = {}) {
  const edge = options.edge ?? false;
  const additionalExcludedPaths = options.excludedPaths ?? [];
  let resolvedConfig;
  let isProductionSsrBuild = false;
  let currentCommand;
  let isHydrogenSite = false;
  let userServerFile;
  return {
    name: "vite-plugin-netlify-react-router",
    config(_config, { command, isSsrBuild }) {
      currentCommand = command;
      isProductionSsrBuild = isSsrBuild === true && command === "build";
      if (isProductionSsrBuild) {
        const functionHandlerInput = {
          [FUNCTION_HANDLER_CHUNK]: FUNCTION_HANDLER_MODULE_ID
        };
        const mergedInput = mergeRollupInput(_config.build?.rollupOptions?.input, functionHandlerInput);
        const configChanges = {
          build: {
            rollupOptions: {
              input: mergedInput,
              output: {
                // NOTE: must use function syntax here to work around Shopify CLI reading
                // the config value literally (i.e. trying to stat `[name].js` as a filename).
                entryFileNames: () => "[name].js"
              }
            }
          },
          // Additional config needed for Edge Functions if enabled
          ...edge ? {
            ssr: {
              target: "webworker",
              // Bundle everything except Node.js built-ins (which are supported but must use the `node:` prefix):
              // https://docs.netlify.com/build/edge-functions/api/#runtime-environment
              noExternal: /^(?!node:).*$/,
              resolve: {
                conditions: ["worker", "deno", "browser"]
              }
            }
          } : {}
        };
        return configChanges;
      }
    },
    async resolveId(source, importer, options2) {
      if (isHydrogenSite && edge) {
        if (source === FUNCTION_HANDLER_MODULE_ID || source === SERVER_ENTRY_MODULE_ID) {
          return;
        }
      }
      if (source === FUNCTION_HANDLER_MODULE_ID) {
        return RESOLVED_FUNCTION_HANDLER_MODULE_ID;
      }
      if (source === SERVER_ENTRY_MODULE_ID && edge) {
        if (currentCommand === "serve") {
          const reactRouterDev = await this.resolve("@react-router/dev/config", importer, options2);
          if (!reactRouterDev) {
            throw new Error("The @react-router/dev package is required for local development. Please install it.");
          }
          return (0, import_node_path2.resolve)((0, import_node_path2.dirname)(reactRouterDev.id), "config/defaults/entry.server.node.tsx");
        }
        return this.resolve("@netlify/vite-plugin-react-router/entry.server.edge", importer, options2);
      }
    },
    // See https://vitejs.dev/guide/api-plugin#virtual-modules-convention.
    load(id) {
      if (id === RESOLVED_FUNCTION_HANDLER_MODULE_ID) {
        return edge ? EDGE_FUNCTION_HANDLER : FUNCTION_HANDLER;
      }
    },
    configResolved: {
      order: "pre",
      async handler(config) {
        resolvedConfig = config;
        isHydrogenSite = config.plugins.some((plugin) => plugin.name === "hydrogen:main");
        if (isHydrogenSite && edge) {
          userServerFile = await findUserEdgeFunctionHandlerFile(config.root);
          if (isProductionSsrBuild) {
            if (config.build?.rollupOptions?.input && typeof config.build.rollupOptions.input === "object" && !Array.isArray(config.build.rollupOptions.input)) {
              config.build.rollupOptions.input[FUNCTION_HANDLER_CHUNK] = userServerFile;
            }
          }
        }
      }
    },
    // In dev, Hydrogen sites need their server.ts to be loaded and called for each request so that
    // it can provide `getLoadContext` (storefront, cart, session, etc.) to React Router's request
    // handler. Without this, React Router's dev middleware would handle SSR with no load context.
    configureServer: {
      order: "pre",
      handler(viteDevServer) {
        if (!isHydrogenSite || !edge) return;
        if (!userServerFile) {
          viteDevServer.config.logger.warn(
            "Hydrogen site detected but no server.ts found. Dev SSR will fall through to React Router defaults."
          );
          return;
        }
        const serverEntryFile = userServerFile;
        return () => {
          viteDevServer.middlewares.use(async (req, res, next) => {
            try {
              const serverModule = await viteDevServer.ssrLoadModule((0, import_node_path2.join)(viteDevServer.config.root, serverEntryFile));
              const handler = serverModule.default;
              req.url = req.originalUrl ?? req.url;
              const request = (0, import_node_fetch_server.createRequest)(req, res);
              const netlifyContext = { waitUntil: () => {
              }, ...globalThis.Netlify?.context };
              const response = await handler(request, netlifyContext);
              if (response) {
                await (0, import_node_fetch_server.sendResponse)(res, response);
              } else {
                next();
              }
            } catch (error) {
              next(error);
            }
          });
        };
      }
    },
    // See https://rollupjs.org/plugin-development/#writebundle.
    async writeBundle() {
      if (isProductionSsrBuild) {
        const handlerPath = (0, import_node_path2.join)(resolvedConfig.build.outDir, `${FUNCTION_HANDLER_CHUNK}.js`);
        if (edge) {
          const clientDir = (0, import_node_path2.join)(resolvedConfig.build.outDir, "..", "client");
          const clientFiles = await (0, import_tinyglobby.glob)("**/*", {
            cwd: clientDir,
            // We can't exclude entire directories because there could be `foo/bar.baz` in the
            // client dir and a `/foo` route handled by the server function.
            onlyFiles: true,
            dot: true
          });
          const excludedPath = ["/.netlify/*", ...clientFiles.map((file) => `/${file}`), ...additionalExcludedPaths];
          const edgeFunctionsDir = (0, import_node_path2.join)(resolvedConfig.root, NETLIFY_EDGE_FUNCTIONS_DIR);
          await (0, import_promises.mkdir)(edgeFunctionsDir, { recursive: true });
          const relativeHandlerPath = toPosixPath((0, import_node_path2.relative)(edgeFunctionsDir, handlerPath));
          await (0, import_promises.writeFile)(
            (0, import_node_path2.join)(edgeFunctionsDir, FUNCTION_FILENAME),
            generateEdgeFunction(relativeHandlerPath, excludedPath)
          );
        } else {
          const functionsDir = (0, import_node_path2.join)(resolvedConfig.root, NETLIFY_FUNCTIONS_DIR);
          await (0, import_promises.mkdir)(functionsDir, { recursive: true });
          const relativeHandlerPath = toPosixPath((0, import_node_path2.relative)(functionsDir, handlerPath));
          const excludedPath = ["/.netlify/*", ...additionalExcludedPaths];
          await (0, import_promises.writeFile)(
            (0, import_node_path2.join)(functionsDir, FUNCTION_FILENAME),
            generateNetlifyFunction(relativeHandlerPath, excludedPath)
          );
        }
      }
    }
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createRequestHandler,
  netlifyRouterContext
});
